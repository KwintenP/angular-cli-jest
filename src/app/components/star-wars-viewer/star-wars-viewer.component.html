<p>
  star-wars-viewer works!
</p>

<form [formGroup]="idFormGroup" (ngSubmit)="getCharacter()">
  <label for="starWarsId">StarWars Id</label>
  <input type="text" id="starWarsId" formControlName="id"/>
  <button type="submit">Get character</button>
</form>

{{character$Â | async | json}}
