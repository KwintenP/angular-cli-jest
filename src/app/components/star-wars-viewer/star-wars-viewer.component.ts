import {Component, OnInit} from '@angular/core';
import {FormBuilder} from '@angular/forms';
import {StarWarsService} from '../../services/star-wars.service';
import {NotificationService} from '../../services/notification.service';

@Component({
  selector: 'app-star-wars-viewer',
  templateUrl: './star-wars-viewer.component.html',
  styleUrls: ['./star-wars-viewer.component.css']
})
export class StarWarsViewerComponent implements OnInit {
  idFormGroup;
  character$;

  constructor(private formBuilder: FormBuilder,
              private starwarsService: StarWarsService,
              private notificationService: NotificationService) {
  }

  ngOnInit() {
    this.idFormGroup = this.formBuilder.group({
      id: ''
    });
  }

  getCharacter() {
    this.character$ = this.starwarsService.getStarWarsCharacter(this.idFormGroup.value.id)
      .do(_ => this.notificationService.showSuccess('Fetched data correctly'));
  }
}
