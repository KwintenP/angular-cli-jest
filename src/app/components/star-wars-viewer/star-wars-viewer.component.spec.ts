import {async, ComponentFixture, TestBed} from '@angular/core/testing';

import {StarWarsViewerComponent} from './star-wars-viewer.component';
import {FormBuilder, FormControl, FormGroup} from '@angular/forms';
import {Observable} from 'rxjs/Observable';
import {NO_ERRORS_SCHEMA} from '@angular/core';
import {NotificationService} from '../../services/notification.service';
import {StarWarsService} from '../../services/star-wars.service';

describe('StarWarsViewerComponent', () => {
  let componentUT: StarWarsViewerComponent;
  let formBuilderMock;
  let starWarsServiceMock;
  let notificationServiceMock;

  beforeEach(() => {
    formBuilderMock = {
      group: jest.fn()
    };
    starWarsServiceMock = {
      getStarWarsCharacter: jest.fn()
    };
    notificationServiceMock = {
      showSuccess: jest.fn()
    };
  });

  describe('snapshot testing', () => {
    let component: StarWarsViewerComponent;
    let fixture: ComponentFixture<StarWarsViewerComponent>;

    beforeEach(async(() => {
        TestBed.configureTestingModule({
          declarations: [StarWarsViewerComponent],
          providers: [
            {provide: FormBuilder, useValue: formBuilderMock},
            {provide: NotificationService, useValue: notificationServiceMock},
            {provide: StarWarsService, useValue: starWarsServiceMock}
          ],
          schemas: [NO_ERRORS_SCHEMA]
        }).compileComponents();
      })
    );

    beforeEach(() => {
      let formGroupMock = new FormGroup({
        id: new FormControl()
      });
      formBuilderMock.group.mockReturnValue(formGroupMock);

      fixture = TestBed.createComponent(StarWarsViewerComponent);
      component = fixture.componentInstance;
      fixture.detectChanges();
    });

    it('should be created', () => {
      expect(fixture).toMatchSnapshot();
    });

    it('should list the characters when the data is fetched', () => {
      let character = {name: 'Luke Skywalker'};
      let returnValue = Observable.of(character);
      starWarsServiceMock.getStarWarsCharacter.mockReturnValue(returnValue);

      component.getCharacter();
      fixture.detectChanges();

      expect(fixture).toMatchSnapshot();
    });
  });

  describe('unit tests', () => {
    beforeEach(() => {
      componentUT = new StarWarsViewerComponent(
        formBuilderMock,
        starWarsServiceMock,
        notificationServiceMock
      );
    });

    describe('on ngOnInit', () => {
      it('should init the formGroup', () => {
        let formGroupMock = new FormGroup({
          id: new FormControl()
        });
        formBuilderMock.group.mockReturnValue(formGroupMock);

        componentUT.ngOnInit();

        expect(componentUT.idFormGroup).toBe(formGroupMock);
        expect(formBuilderMock.group).toHaveBeenCalled();
        expect(formBuilderMock.group.mock.calls.length).toBe(1);
        expect(formBuilderMock.group.mock.calls[0][0]).toEqual({
          id: ''
        });
      });
    });

    describe('on getCharacter', () => {
      it('should fetch a char$ from the starWarsService', () => {
        componentUT.idFormGroup = new FormGroup({
          id: new FormControl()
        });

        let character = {name: 'Luke Skywalker'};
        let returnValue = Observable.of(character);
        starWarsServiceMock.getStarWarsCharacter.mockReturnValue(returnValue);

        componentUT.getCharacter();

        componentUT.character$.subscribe(
          (val) => {
            expect(val).toBe(character)
            expect(notificationServiceMock.showSuccess).toHaveBeenCalled();
          }
        );
      });
    });
  });
})
;
