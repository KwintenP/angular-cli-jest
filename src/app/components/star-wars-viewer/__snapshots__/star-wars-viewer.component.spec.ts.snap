// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarWarsViewerComponent snapshot testing should be created 1`] = `
<app-star-wars-viewer
  formBuilder={[Function Object]}
  idFormGroup={[Function FormGroup]}
  notificationService={[Function Object]}
  starwarsService={[Function Object]}
>
  <p>
    star-wars-viewer works!
  </p><form>
    <label
      for="starWarsId"
    >
      StarWars Id
    </label>
    <input
      formcontrolname="id"
      id="starWarsId"
      type="text"
    />
    <button
      type="submit"
    >
      Get character
    </button>
  </form>null
</app-star-wars-viewer>
`;

exports[`StarWarsViewerComponent snapshot testing should list the characters when the data is fetched 1`] = `
<app-star-wars-viewer
  character$={[Function Observable]}
  formBuilder={[Function Object]}
  idFormGroup={[Function FormGroup]}
  notificationService={[Function Object]}
  starwarsService={[Function Object]}
>
  <p>
    star-wars-viewer works!
  </p><form>
    <label
      for="starWarsId"
    >
      StarWars Id
    </label>
    <input
      formcontrolname="id"
      id="starWarsId"
      type="text"
    />
    <button
      type="submit"
    >
      Get character
    </button>
  </form>{ "name": "Luke Skywalker" }
</app-star-wars-viewer>
`;
